README.md
# tslib
A correct and consistent API for dealing with leads, lags, differences, and filling in gaps in time-series and panel data. Available in Pandas and PySpark (via pandas-on-Spark).

In Pandas, importing `tslib` grants access to the `.ts` accessor, allowing for idiomatic creation of lags, leads, and differences with time series and panel data. 

All Pandas functionality works equally with [pandas-on-Spark](https://spark.apache.org/docs/latest/api/python/user_guide/pandas_on_spark/pandas_pyspark.html) DataFrames.

See complete documentation [here](https://jacgoldsm.github.io/tslib/).

## Installation
`tslib` is not currently on PyPI. Install from GitHub with:

`pip install tslib@git+https://github.com/jacgoldsm/tslib`

## Using with PySpark
To use with the normal PySpark DataFrames, sandwich your `tslib` code as follows:

```python
my_pandas_on_spark = my_spark_df.pandas_api()
# Use `tslib` Pandas APIs as normal here
my_new_spark_df = my_pandas_on_spark.to_spark()
```

## Getting Started—Time-Series Data
```python
import pandas as pd
from tslib.pandas_api import TimeOpts

# Define our Data Frame. `tslib` works with dates stored as numbers or as Pandas dates.
cookies = pd.DataFrame(
    {
        "year": [2000, 2001, 2002, 2003, 2008],
        "favorite": [
            "Chocolate Chip",
            "Chocolate Chip",
            "Oatmeal Raisin",
            "Sugar",
            "M&M",
        ],
        "n": [10, 20, 15, 12, 40],
    }
)
print(cookies)
    year       favorite   n
0  2000  Chocolate Chip  10
1  2001  Chocolate Chip  20
2  2002  Oatmeal Raisin  15
3  2003           Sugar  12
4  2008             M&M  40
# Define our time series arguments. 
# Set the time-series column, frequency, and start of the time-series.
cookies_args = TimeOpts(ts_column="year", freq=1, start=1999)
# define our time-series DataFrame
cookies_ts = cookies.ts(cookies_args)
# create a DataFrame with all the gaps in the time-series filled in
print(cookies_ts.tsfill())
    year       favorite     n
0  1999             NaN  <NA>
1  2000  Chocolate Chip    10
2  2001  Chocolate Chip    20
3  2002  Oatmeal Raisin    15
4  2003           Sugar    12
5  2004             NaN  <NA>
6  2005             NaN  <NA>
7  2006             NaN  <NA>
8  2007             NaN  <NA>
9  2008             M&M    40
# create a DataFrame with the lagged value of `favorite`. 
# Note that lag respects gaps in the data,
# so year 2008 has no lag since there's no value for 2007
print(cookies_ts.with_lag("previous_favorite",column="favorite"))
    year       favorite   n previous_favorite
0  2000  Chocolate Chip  10               NaN
1  2001  Chocolate Chip  20    Chocolate Chip
2  2002  Oatmeal Raisin  15    Chocolate Chip
3  2003           Sugar  12    Oatmeal Raisin
4  2008             M&M  40               NaN
# the same, but with a lead
print(cookies_ts.with_lead("next_favorite",column="favorite"))
   year        favorite   n   next_favorite
0  2000  Chocolate Chip  10  Chocolate Chip
1  2001  Chocolate Chip  20  Oatmeal Raisin
2  2002  Oatmeal Raisin  15           Sugar
3  2003           Sugar  12             NaN
4  2008             M&M  40             NaN
# the same, but with differencing
print(cookies_ts.with_difference("change_in_panelists",column="n"))
    year       favorite   n  change_in_panelists
0  2000  Chocolate Chip  10                  NaN
1  2001  Chocolate Chip  20                 10.0
2  2002  Oatmeal Raisin  15                 -5.0
3  2003           Sugar  12                 -3.0
4  2008             M&M  40                  NaN
```

## Getting Started: Panel Data
```python
import pandas as pd
from tslib.pandas_api import TimeOpts

# Define our Data Frame. `tslib` works with dates stored as numbers or as Pandas dates.
dates = [
    "2000-01-01",
    "2000-01-01",
    "2000-01-01",
    "2000-02-01",
    "2000-02-01",
    "2000-02-01",
    "2000-04-01",
    "2000-04-01",
    "2000-05-01",
    "2000-05-01",
    "2000-09-01",
    "2000-09-01",
    "2000-09-01",
]

panel = pd.DataFrame(
    {
        "id": [1, 2, 3, 1, 2, 3, 1, 2, 1, 3, 1, 2, 3],
        "date": [pd.to_datetime(date) for date in dates],
        "sex": ["m", "f", "f", "m", "f", "f", "m", "f", "m", "f", "m", "f", "f"],
        "credit_score": [
            750,
            820,
            640,
            760,
            810,
            670,
            740,
            840,
            745,
            620,
            780,
            800,
            630,
        ],
    }
)
print(panel)
    id       date sex  credit_score
0    1 2000-01-01   m           750
1    2 2000-01-01   f           820
2    3 2000-01-01   f           640
3    1 2000-02-01   m           760
4    2 2000-02-01   f           810
5    3 2000-02-01   f           670
6    1 2000-04-01   m           740
7    2 2000-04-01   f           840
8    1 2000-05-01   m           745
9    3 2000-05-01   f           620
10   1 2000-09-01   m           780
11   2 2000-09-01   f           800
12   3 2000-09-01   f           630
# define our Time-Series options
panel_args = TimeOpts(
    ts_column="date",
    panel_column=panel["id"],
    start="1999-12-01",
    end="2000-10-01",
    freq="1m",
)

# set up our time-series DataFrame
panel_ts = panel.ts(panel_args)

# fill in our complete panel
print(panel_ts.tsfill())
    id       date  sex  credit_score
0    1 1999-12-01  NaN          <NA>
1    1 2000-01-01    m           750
2    1 2000-02-01    m           760
3    1 2000-03-01  NaN          <NA>
4    1 2000-04-01    m           740
5    1 2000-05-01    m           745
6    1 2000-06-01  NaN          <NA>
7    1 2000-07-01  NaN          <NA>
8    1 2000-08-01  NaN          <NA>
9    1 2000-09-01    m           780
10   1 2000-10-01  NaN          <NA>
11   2 1999-12-01  NaN          <NA>
12   2 2000-01-01    f           820
13   2 2000-02-01    f           810
14   2 2000-03-01  NaN          <NA>
15   2 2000-04-01    f           840
16   2 2000-05-01  NaN          <NA>
17   2 2000-06-01  NaN          <NA>
18   2 2000-07-01  NaN          <NA>
19   2 2000-08-01  NaN          <NA>
20   2 2000-09-01    f           800
21   2 2000-10-01  NaN          <NA>
22   3 1999-12-01  NaN          <NA>
23   3 2000-01-01    f           640
24   3 2000-02-01    f           670
25   3 2000-03-01  NaN          <NA>
26   3 2000-04-01  NaN          <NA>
27   3 2000-05-01    f           620
28   3 2000-06-01  NaN          <NA>
29   3 2000-07-01  NaN          <NA>
30   3 2000-08-01  NaN          <NA>
31   3 2000-09-01    f           630
32   3 2000-10-01  NaN          <NA>
# create our lagged data with lag of 2, with gaps preserved
print(panel_ts.with_lag("lag_credit",column="credit_score",back=2))
    id       date sex  credit_score  lag_credit
0    1 2000-01-01   m           750        <NA>
1    2 2000-01-01   f           820        <NA>
2    3 2000-01-01   f           640        <NA>
3    1 2000-02-01   m           760        <NA>
4    2 2000-02-01   f           810        <NA>
5    3 2000-02-01   f           670        <NA>
6    1 2000-04-01   m           740         760
7    2 2000-04-01   f           840         810
8    1 2000-05-01   m           745        <NA>
9    3 2000-05-01   f           620        <NA>
10   1 2000-09-01   m           780        <NA>
11   2 2000-09-01   f           800        <NA>
12   3 2000-09-01   f           630        <NA>
# the same, but with a lead
print(panel_ts.with_lead("lag_credit",column="credit_score",forward=2))
    id       date sex  credit_score  lead_credit
0    1 2000-01-01   m           750         <NA>
1    2 2000-01-01   f           820         <NA>
2    3 2000-01-01   f           640         <NA>
3    1 2000-02-01   m           760          740
4    2 2000-02-01   f           810          840
5    3 2000-02-01   f           670         <NA>
6    1 2000-04-01   m           740         <NA>
7    2 2000-04-01   f           840         <NA>
8    1 2000-05-01   m           745         <NA>
9    3 2000-05-01   f           620         <NA>
10   1 2000-09-01   m           780         <NA>
11   2 2000-09-01   f           800         <NA>
12   3 2000-09-01   f           630         <NA>
# the same, but with differencing
print(panel_ts.with_difference(name="credit_change",column="credit_score",back=2))
    id       date sex  credit_score  credit_change
0    1 2000-01-01   m           750           <NA>
1    2 2000-01-01   f           820           <NA>
2    3 2000-01-01   f           640           <NA>
3    1 2000-02-01   m           760           <NA>
4    2 2000-02-01   f           810           <NA>
5    3 2000-02-01   f           670           <NA>
6    1 2000-04-01   m           740            -20
7    2 2000-04-01   f           840             30
8    1 2000-05-01   m           745           <NA>
9    3 2000-05-01   f           620           <NA>
10   1 2000-09-01   m           780           <NA>
11   2 2000-09-01   f           800           <NA>
12   3 2000-09-01   f           630           <NA>
```

## Contributing

Before making a pull request, run the tests to ensure that none of them are broken. You can do 
this with the following code:
```bash
python3 -m pip install pytest
python3 -m pytest
```



CHANGELOG.md
# Changelog

## Unrealeased

- ...

tests/test_ts.py
from __future__ import annotations

from tslib.pandas_api import TimeOpts
import pandas as pd

from pandas.testing import assert_frame_equal

cookies = pd.DataFrame(
    {
        "year": [2000, 2001, 2002, 2003, 2008],
        "favorite": [
            "Chocolate Chip",
            "Chocolate Chip",
            "Oatmeal Raisin",
            "Sugar",
            "M&M",
        ],
        "n": [10, 20, 15, 12, 40],
    }
)
cookies_args = TimeOpts(ts_column="year", freq=1, start=1999)
cookies_ts = cookies.ts(cookies_args)
cookies_full = cookies_ts.tsfill()
cookies_with_lag = cookies_ts.with_lag("previous_favorite", column="favorite")
cookies_with_lead = cookies_ts.with_lead("next_favorite", column="favorite")
cookies_with_diff = cookies_ts.with_difference("change_in_panelists", column="n")
print(cookies)
print(cookies_full)
print(cookies_with_lag)
print(cookies_with_lead)
print(cookies_with_diff)

cookies_date = pd.DataFrame(
    {
        "year": (
            pd.to_datetime(i)
            for i in [
                "2000-01-01",
                "2001-01-01",
                "2002-01-01",
                "2003-01-01",
                "2008-01-01",
            ]
        ),
        "favorite": [
            "Chocolate Chip",
            "Chocolate Chip",
            "Oatmeal Raisin",
            "Sugar",
            "M&M",
        ],
        "n": [10, 20, 15, 12, 40],
    }
)
cookies_date_args = TimeOpts(ts_column="year", freq="Y", start="1999-01-01")
cookies_date_ts = cookies_date.ts(cookies_date_args)
cookies_date_full = cookies_date_ts.tsfill()
cookies_date_with_lag = cookies_date_ts.with_lag("previous_favorite", column="favorite")
cookies_date_with_lead = cookies_date_ts.with_lead("next_favorite", column="favorite")
cookies_date_with_diff = cookies_date_ts.with_difference(
    "change_in_panelists", column="n"
)


expected_filled = pd.DataFrame(
    {
        "year": [1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008],
        "favorite": [
            None,
            "Chocolate Chip",
            "Chocolate Chip",
            "Oatmeal Raisin",
            "Sugar",
            None,
            None,
            None,
            None,
            "M&M",
        ],
        "n": [None, 10, 20, 15, 12, None, None, None, None, 40],
    }
)

expected_filled_date = expected_filled.copy()
expected_filled_date.year = [pd.to_datetime(f"{i}-01-01") for i in expected_filled.year]

expected_lag = cookies.copy()
expected_lag["previous_favorite"] = [
    None,
    "Chocolate Chip",
    "Chocolate Chip",
    "Oatmeal Raisin",
    None,
]

expected_lag_date = cookies_date.copy()
expected_lag_date["previous_favorite"] = [
    None,
    "Chocolate Chip",
    "Chocolate Chip",
    "Oatmeal Raisin",
    None,
]

expected_lead = cookies.copy()
expected_lead["next_favorite"] = [
    "Chocolate Chip",
    "Oatmeal Raisin",
    "Sugar",
    None,
    None,
]

expected_lead_date = cookies_date.copy()
expected_lead_date["next_favorite"] = [
    "Chocolate Chip",
    "Oatmeal Raisin",
    "Sugar",
    None,
    None,
]

expected_diff = cookies.copy()
expected_diff["change_in_panelists"] = [None, 10, -5, -3, None]

expected_diff_date = cookies_date.copy()
expected_diff_date["change_in_panelists"] = [None, 10, -5, -3, None]


import pyspark.pandas as ps


def compares_equal(
    df1: pd.DataFrame | ps.DataFrame,
    df2: pd.DataFrame | ps.DataFrame,
    check_dtype: bool = False,
) -> bool:
    if isinstance(df1, ps.DataFrame):
        df1 = df1.to_pandas()
        df2 = df2.to_pandas()
    try:
        assert_frame_equal(df1, df2, check_dtype=check_dtype, check_index_type=False)
        return True
    except AssertionError:
        return False

print(cookies_full)

def test_cookies():
    assert compares_equal(cookies_date_full, expected_filled_date)
    assert compares_equal(cookies_date_with_lag, expected_lag_date)
    assert compares_equal(cookies_date_with_lead, expected_lead_date)
    assert compares_equal(cookies_date_with_diff, expected_diff_date)

    assert compares_equal(cookies_full, expected_filled)
    assert compares_equal(cookies_with_lag, expected_lag)
    assert compares_equal(cookies_with_lead, expected_lead)
    assert compares_equal(cookies_with_diff, expected_diff)


dataframes = [
    (name, df) for (name, df) in globals().items() if isinstance(df, pd.DataFrame)
]
for tup in dataframes:
    name, df = tup
    globals()["ps_" + name] = ps.from_pandas(df)


def test_cookies_ps():
    assert compares_equal(ps_cookies_date_full, ps_expected_filled_date)
    assert compares_equal(ps_cookies_date_with_lag, ps_expected_lag_date)
    assert compares_equal(ps_cookies_date_with_lead, ps_expected_lead_date)
    assert compares_equal(ps_cookies_date_with_diff, ps_expected_diff_date)

    assert compares_equal(ps_cookies_full, ps_expected_filled)
    assert compares_equal(ps_cookies_with_lag, ps_expected_lag)
    assert compares_equal(ps_cookies_with_lead, ps_expected_lead)
    assert compares_equal(ps_cookies_with_diff, ps_expected_diff)


tslib/pandas_api.py
from __future__ import annotations

import pandas as pd
from typing import Any, Iterable, TYPE_CHECKING
from dataclasses import dataclass
from datetime import datetime

from tslib.pandas_utils import (
    _is_acceptable_type,
    _is_date,
    _is_numeric,
    _is_numeric_scalar,
    TimeDict,
)

try:
    from pyspark.pandas.extensions import (
        register_dataframe_accessor as pyspark_register_dataframe_accessor,
    )
except ImportError:
    pass

if TYPE_CHECKING:
    import pyspark.pandas as ps
    import numpy as np


@dataclass
class TimeOpts:
    ts_column: str | pd.Series | ps.Series
    panel_column: str | pd.Series | ps.Series | None = None
    freq: int | float | str | pd.offsets.DateOffset | None = None
    start: int | float | str | pd.Timestamp | datetime | np.datetime64 | None = None
    end: int | float | str | pd.Timestamp | datetime | np.datetime64 | None = None

    def assign(self, inplace=False, **kwargs) -> TimeOpts | None:
        import copy

        if not inplace:
            out = copy.copy(self)
        else:
            out = None
        for kwarg in kwargs.keys():
            if kwarg not in ("ts_column", "panel_column", "freq", "start", "end"):
                raise ValueError(f"Invalid option: {kwarg}")
            if inplace:
                setattr(self, kwarg, kwargs[kwarg])
            else:
                setattr(out, kwarg, kwargs[kwarg])
        return out

    @staticmethod
    def _extract_opt(opts: TimeOpts | dict, opt: str):
        if isinstance(opts, TimeOpts):
            return getattr(opts, opt)
        else:
            return opts[opt]

    @staticmethod
    def _extract_opts(opts: TimeOpts | dict):
        if isinstance(opts, TimeOpts):
            return (opts.ts_column, opts.panel_column, opts.freq, opts.start, opts.end)
        else:
            out = TimeOpts(
                opts["ts_column"],
                opts.get("panel_column", None),
                opts.get("freq", None),
                opts.get("start", None),
                opts.get("end", None),
            )
            return (out.ts_column, out.panel_column, out.freq, out.start, out.end)


def _maybe_pyspark(name):
    def deco(cls):
        try:
            return pyspark_register_dataframe_accessor(cls)
        except:
            return cls
    return deco
        
@_maybe_pyspark("ts")
@pd.api.extensions.register_dataframe_accessor("ts")
class TSAccessor:
    def __init__(self, obj: pd.DataFrame | ps.DataFrame):
        try:
            import pyspark.pandas as ps

            if isinstance(obj, ps.DataFrame):
                self._engine = ps
            else:
                self._engine = pd
        except ImportError:
            self._engine = pd

        self._obj = obj

    def __call__(self, opts: TimeOpts | dict) -> TSAccessor:
        self._opts = opts
        return self

    def tsset(
        self,
        ts_args: TimeOpts | dict,
    ) -> TimeDict:
        data = self._obj.copy()
        ts_column, panel_column, freq, start, end = TimeOpts._extract_opts(ts_args)
        out_dict = TimeDict()
        if isinstance(ts_column, (str, bytes)):
            ts_column_name = ts_column
            ts_column = self._obj[ts_column]
            if len(ts_column) < 2:
                raise ValueError("A time series needs at least two records.")
        else:
            ts_column_name: str = (
                str(ts_column.name) if ts_column.name is not None else "time_series"
            )
        out_dict.ts_column_name = ts_column_name
        if not _is_acceptable_type(ts_column):
            raise Exception(f"{ts_column.dtype} not an acceptable time-series type")
        if len(ts_column.index) != ts_column.nunique() and panel_column is None:
            raise Exception("Time-series column cannot contain duplicates")
        if _is_numeric(ts_column) and not _is_numeric_scalar(freq):
            raise Exception(
                "For a numeric time-series variable" "Frequency must be a number."
            )
        data.index = self._engine.RangeIndex(start=0,stop=len(data.index))
        is_date = _is_date(ts_column)
        if is_date:
            out_dict.freq = pd.tseries.frequencies.to_offset(freq)  # type: ignore
            assert not isinstance(start, (int, float))
            assert not isinstance(end, (int, float))
        else:
            assert isinstance(freq, (int, float))
            out_dict.freq = freq

        if start is not None:
            if is_date:
                start = pd.to_datetime(start)
        else:
            start = ts_column.min()

        if end is not None:
            if is_date:
                end = pd.to_datetime(end)
        else:
            end = ts_column.max()

        if is_date:
            complete_time_series = self._engine.period_range(
                start=start, end=end, freq=out_dict.freq
            ).to_timestamp()  # type: ignore
        else:
            # note that `end` should be inclusive whereas RangeIndex is top-exclusive
            complete_time_series = self._engine.RangeIndex(
                start=start,stop=end + freq, step=out_dict.freq
            )
        if set(ts_column[(ts_column >= start) & (ts_column <= end)]) - set(
            complete_time_series
        ):
            raise Exception("Time series structure doesn't match frequency specified.")

        if panel_column is not None:
            if isinstance(panel_column, (str, bytes)):
                panel_column_name = panel_column
                panel_column = self._obj[panel_column]
            else:
                panel_column_name: str | None = (
                    str(panel_column.name) if panel_column.name is not None else "panel"
                )
        else:
            panel_column_name = None

        out_dict.complete_time_series = complete_time_series
        out_dict.data = data
        out_dict.start, out_dict.end = start, end
        out_dict.is_date = is_date
        out_dict.panel_column_name = panel_column_name
        out_dict.data.sort_values(ts_column_name, inplace=True)
        out_dict.engine = self._engine
        return out_dict

    def tsfill(
        self,
        *,
        fill_value: Any = None,
        method: str | None = None,
        sentinel: str | None = None,
        keep_index: bool = False,
        avoid_float_casts: bool = True,
        opts_replacement: TimeOpts | dict | None = None,
    ) -> pd.DataFrame | ps.DataFrame:
        """
        Fill in holes in time-series or panel data

        Args:
            fill_value: Value to fill in NAs passed to `pandas.DataFrame.reindex`
            method: Method of filling in NAs passed to `pandas.DataFrame.reindex`
            sentinel: If not None, the name of a column indicating if a row was
                present in the original data (True) or was filled in by `tsfill`
                (False)
            keep_index: If True, the index of the data returned will be the index of
                the original data with null values for filled-in observations
            avoid_float_casts: Use Pandas nullable dtypes to avoid casting integer columns
                to float when NAs are filled in.
            opts_replacement: Replace Arguments for the time-series structure of the data.
                Defaults to the existing TimeOpts arguments from `ts()`


        Returns:
            Pandas DataFrame or pandas-on-Spark DataFrame

        Examples:
            >>> from tslib.pandas_api import TimeOpts
            >>> import pandas as pd
            >>> cookies = pd.DataFrame(
            ...   {
            ...        "year": [2000, 2001, 2002, 2003, 2008],
            ...        "favorite": [
            ...         "Chocolate Chip",
            ...         "Chocolate Chip",
            ...         "Oatmeal Raisin",
            ...         "Sugar",
            ...         "M&M",
            ...    ],
            ...    "n": [10, 20, 15, 12, 40],
            ...   }
            ... )
            >>> cookies
               year        favorite   n
            0  2000  Chocolate Chip  10
            1  2001  Chocolate Chip  20
            2  2002  Oatmeal Raisin  15
            3  2003           Sugar  12
            4  2008             M&M  40
            >>> cookies_args = TimeOpts(ts_column="year", freq=1, start=1999)
            >>> cookies_ts = cookies.ts(cookies_args)
            >>> cookies_full = cookies_ts.tsfill()
            >>> cookies_full
               year        favorite     n
            0  1999             NaN  <NA>
            1  2000  Chocolate Chip    10
            2  2001  Chocolate Chip    20
            3  2002  Oatmeal Raisin    15
            4  2003           Sugar    12
            5  2004             NaN  <NA>
            6  2005             NaN  <NA>
            7  2006             NaN  <NA>
            8  2007             NaN  <NA>
            9  2008             M&M    40

        """
        if opts_replacement is not None:
            ts_args = opts_replacement
        else:
            ts_args = self._opts
        is_panel = TimeOpts._extract_opt(ts_args, "panel_column") is not None
        ts = self.tsset(ts_args=ts_args)
        assert ts.data is not None
        out = ts.data.copy()
        if sentinel is not None:
            sentinel = str(sentinel)
            out[sentinel] = True
        if keep_index:
            out["__index__"] = out.index
        if avoid_float_casts:
            # prevent ints from being turned into floats because of NAs
            out = out.convert_dtypes(
                convert_string=False, infer_objects=False, convert_floating=False  # type: ignore
            )

        assert ts.ts_column_name is not None
        if not is_panel:
            assert isinstance(ts.complete_time_series, Iterable)
            out.index = out[ts.ts_column_name]  # type: ignore
            out = out.reindex(
                ts.complete_time_series, method=method, fill_value=fill_value
            )
            out[ts.ts_column_name] = ts.complete_time_series

        else:
            assert ts.panel_column_name is not None
            out.index = out[ts.ts_column_name]  # type: ignore
            out_grouped = out.groupby(ts.panel_column_name)
            new_groups: list[None | pd.DataFrame] = [
                None for _ in range(len(out_grouped))
            ]
            for i, (key, _) in zip(range(len(out_grouped)), out_grouped.groups.items()):
                subset = out.loc[out[ts.panel_column_name] == key]
                new_groups[i] = subset.reindex(
                    ts.complete_time_series, method=method, fill_value=fill_value
                )
                new_groups[i][ts.ts_column_name] = ts.complete_time_series
                new_groups[i][ts.panel_column_name] = key
            out = ts.engine.concat(new_groups, axis=0)  # type: ignore
        if sentinel is not None:
            out[sentinel] = out[sentinel].fillna(False)
        if keep_index:
            out.index =out["__index__"]
            out.index.name = None
            out.drop("__index__", inplace=True, axis=1)
        else:
            out.index = ts.engine.RangeIndex(0,len(out.index))
        return out

    def with_lag(
        self,
        col_name: str,
        column: str | pd.Series,
        back: int = 1,
        *,
        opts_replacement: TimeOpts | dict | None = None,
    ) -> pd.DataFrame | ps.DataFrame:
        """
        Add a lag column to a Pandas DataFrame

        Args:
            col_name: What to name the lag column
            column: Column to take the lag of
            back: How many records to go back. Negative values are "leads"
            opts_replacement: Replace Arguments for the time-series structure of the data.
                Defaults to the existing TimeOpts arguments from `ts()`

        Returns:
            Pandas DataFrame or pandas-on-Spark DataFrame

        Examples:
            >>> from tslib.pandas_api import TimeOpts
            >>> import pandas as pd
            >>> cookies = pd.DataFrame(
            ...   {
            ...        "year": [2000, 2001, 2002, 2003, 2008],
            ...        "favorite": [
            ...         "Chocolate Chip",
            ...         "Chocolate Chip",
            ...         "Oatmeal Raisin",
            ...         "Sugar",
            ...         "M&M",
            ...    ],
            ...    "n": [10, 20, 15, 12, 40],
            ...   }
            ... )
            >>> cookies
               year        favorite   n
            0  2000  Chocolate Chip  10
            1  2001  Chocolate Chip  20
            2  2002  Oatmeal Raisin  15
            3  2003           Sugar  12
            4  2008             M&M  40
            >>> cookies_args = TimeOpts(ts_column="year", freq=1, start=1999)
            >>> cookies_ts = cookies.ts(cookies_args)
            >>> cookies_lag = cookies_ts.with_lag("previous_favorite", column="favorite")
            >>> cookies_lag
               year        favorite   n previous_favorite
            0  2000  Chocolate Chip  10               NaN
            1  2001  Chocolate Chip  20    Chocolate Chip
            2  2002  Oatmeal Raisin  15    Chocolate Chip
            3  2003           Sugar  12    Oatmeal Raisin
            4  2008             M&M  40               NaN

        """
        if opts_replacement is not None:
            ts_args = opts_replacement
        else:
            ts_args = self._opts
        ts = self.tsset(ts_args=ts_args)
        assert ts.data is not None
        assert ts.freq is not None
        if isinstance(column, str):
            column_string: str = column
            column = ts.data[column]
        else:
            column_string: str = column.name

        if ts.panel_column_name is None:
            if ts.is_date:
                # pessimisation: we can't reliably predict the next element of the series
                # if the column is a date, so we have to use the "complete" time series.
                # Thus, tsfill the data and then use the "shift"-based lag method.
                # Then filter to the original series.
                out = ts.data.ts(ts_args).tsfill(
                    sentinel="__sentinel__", keep_index=True
                )
                out[col_name] = out[column_string].shift(back)
                out = out[out["__sentinel__"]]
                out.drop(["__sentinel__"], inplace=True, axis=1)
            else:
                assert ts.ts_column_name is not None
                lagged_col = ts.data[ts.ts_column_name] + (ts.freq * back)  # type: ignore
                new = ts.engine.DataFrame(
                    {ts.ts_column_name: lagged_col, col_name: ts.data[column_string]}
                )
                out = ts.data.merge(new, on=ts.ts_column_name, how="left")
        else:
            if ts.is_date:
                out = ts.data.ts(ts_args).tsfill(
                    sentinel="__sentinel__", keep_index=True
                )
                out[col_name] = out.groupby(ts.panel_column_name)[column_string].shift(
                    back
                )
                out = out[out["__sentinel__"]]
                out.drop(["__sentinel__"], inplace=True, axis=1)
            else:
                assert ts.ts_column_name is not None
                lagged_col = ts.data[ts.ts_column_name] + (ts.freq * back)  # type: ignore
                new = ts.engine.DataFrame(
                    {
                        ts.ts_column_name: lagged_col,
                        col_name: ts.data[column_string],
                        ts.panel_column_name: ts.data[ts.panel_column_name],
                    }
                )
                out = ts.data.merge(
                    new, on=[ts.ts_column_name, ts.panel_column_name], how="left"
                )

        out.sort_index(inplace=True)
        return out

    def with_lead(
        self,
        col_name: str,
        column: pd.Series | ps.Series | str,
        forward: int | None = 1,
        opts_replacement: TimeOpts | dict | None = None,
    ) -> pd.DataFrame | ps.DataFrame:
        """
        Add a lead column to a Pandas DataFrame

        Args:
            col_name: What to name the lead column
            column: Column to take the lead of
            forward: How many records to go forward. Negative values are "lags"
            opts_replacement: Replace Arguments for the time-series structure of the data.
                Defaults to the existing TimeOpts arguments from `ts()`

        Returns:
            Pandas DataFrame or pandas-on-Spark DataFrame

        Examples:
            >>> from tslib.pandas_api import TimeOpts
            >>> import pandas as pd
            >>> cookies = pd.DataFrame(
            ...   {
            ...        "year": [2000, 2001, 2002, 2003, 2008],
            ...        "favorite": [
            ...         "Chocolate Chip",
            ...         "Chocolate Chip",
            ...         "Oatmeal Raisin",
            ...         "Sugar",
            ...         "M&M",
            ...    ],
            ...    "n": [10, 20, 15, 12, 40],
            ...   }
            ... )
            >>> cookies
               year        favorite   n
            0  2000  Chocolate Chip  10
            1  2001  Chocolate Chip  20
            2  2002  Oatmeal Raisin  15
            3  2003           Sugar  12
            4  2008             M&M  40
            >>> cookies_args = TimeOpts(ts_column="year", freq=1, start=1999)
            >>> cookies_ts = cookies.ts(cookies_args)
            >>> cookies_lead = cookies_ts.with_lead("next_favorite", column="favorite")
            >>> cookies_lead
               year        favorite   n   next_favorite
            0  2000  Chocolate Chip  10  Chocolate Chip
            1  2001  Chocolate Chip  20  Oatmeal Raisin
            2  2002  Oatmeal Raisin  15           Sugar
            3  2003           Sugar  12             NaN
            4  2008             M&M  40             NaN
        """
        if opts_replacement is not None:
            ts_args = opts_replacement
        else:
            ts_args = self._opts
        return self.with_lag(
            col_name, column, back=-1 * forward, opts_replacement=ts_args
        )

    def with_difference(
        self,
        col_name: str,
        column: str | pd.Series,
        back: int | None = 1,
        *,
        opts_replacement: TimeOpts | dict | None = None,
    ) -> pd.DataFrame | ps.DataFrame:
        """
        Add a difference column to a Pandas DataFrame

        Args:
            col_name: What to name the difference column
            column: Column to take the difference of
            back: How many records to go back to compute difference.
            opts_replacement: Replace Arguments for the time-series structure of the data.
                Defaults to the existing TimeOpts arguments from `ts()`

        Returns:
            Pandas DataFrame or pandas-on-Spark DataFrame

        Examples:
            >>> from tslib.pandas_api import TimeOpts
            >>> import pandas as pd
            >>> cookies = pd.DataFrame(
            ...   {
            ...        "year": [2000, 2001, 2002, 2003, 2008],
            ...        "favorite": [
            ...         "Chocolate Chip",
            ...         "Chocolate Chip",
            ...         "Oatmeal Raisin",
            ...         "Sugar",
            ...         "M&M",
            ...    ],
            ...    "n": [10, 20, 15, 12, 40],
            ...   }
            ... )
            >>> cookies
               year        favorite   n
            0  2000  Chocolate Chip  10
            1  2001  Chocolate Chip  20
            2  2002  Oatmeal Raisin  15
            3  2003           Sugar  12
            4  2008             M&M  40
            >>> cookies_args = TimeOpts(ts_column="year", freq=1, start=1999)
            >>> cookies_ts = cookies.ts(cookies_args)
            >>> cookies_diff = cookies_ts.with_difference("change_in_panelists", column="n")
            >>> cookies_diff
               year        favorite   n  change_in_panelists
            0  2000  Chocolate Chip  10                  NaN
            1  2001  Chocolate Chip  20                 10.0
            2  2002  Oatmeal Raisin  15                 -5.0
            3  2003           Sugar  12                 -3.0
            4  2008             M&M  40                  NaN
        """
        if opts_replacement is not None:
            ts_args = opts_replacement
        else:
            ts_args = self._opts
        lag = self.with_lag(
            "__difference_dummy__", column, back, opts_replacement=ts_args
        )["__difference_dummy__"]
        curr = self._obj[column]
        out = self._obj.copy()
        out[col_name] = curr - lag
        return out

